src/local/butler/reproduce.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""reproduce.py reproduces test cases locally."""
+
+from __future__ import print_function
+
+import os
+import six
+
+from src.python.fuzzing import tests
+from src.python.system import environment
+
+
+class _SimplifiedTestcase(object):
+  """Minimal representation of a test case."""
+
+  def __init__(self, testcase_json):
+    self.crash_state = testcase_json['crash_state']
```

Comment:
This is faulty logic, that came from old reproduce tool. Now we want to serialize/deserialize the whole testcase and job object returned from server (checkout to_dict to serialize and some way to deserialize). there can be numerous fields in testcase that might be get silently used (e.g. metadata or whatever property). Please create a appengine handler that return full entities in some encoded form.

Also _get_testcase() should use that rather than creating testcase itself with few fields. for test, use test_utils.create_generic_testcase.

Reply:
I agree that we should try to handle the serialization/deserialization in a generic way so that we don't have to explicitly list out fields here. I'd like to leave this as-is for this change, but I'll clean this up next.

Right now we're making use of the added job definition field, and I think going forward we'll still want this to be its own class. This was the bare minimum from testcase needed to get this example working, but it wasn't meant to be how we do things long-term. As for using a separate class, since we won't be able to access datastore, it may be useful to be able to override specific fields or functions (e.g. put).

Let me know if this sounds good to you. I think this is the last unresolved point.